### PR TITLE
1241 simple report uptime keeper

### LINKED ISSUE #1241: Design and implement data structures to keep validators report status over the past X requests
Maybe can take a look at how Cosmos maintain node uptime stat. We will want to jail people that don't do the oracle job. Don't need to jail yet with this ticket.

* Implement only data type and design how to store data in storage
 - ReportInfo
 - ValidatorMissedReportBitArray

### PR DESCRIPTION
**Pull request by taobun**
fixed #1241

### CODE REVIEW

File: chain/x/oracle/types/params.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ const (
 
 	// Prepare gas cost value is 100000
 	DefaultPrepareGas = uint64(100000)
+
+	// Reported window size is 100
```
Comment (sorawit) [inline chain/x/oracle/types/params.go]: Change comment

File: chain/x/oracle/keeper/hooks.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
+)
+
+func (k Keeper) AfterValidatorBonded(ctx sdk.Context, _ sdk.ConsAddress, address sdk.ValAddress) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/hooks.go]: Should we remove the second argument?

File: chain/x/oracle/keeper/hooks.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+var _ types.StakingHooks = Hooks{}
+
+// Return the wrapper struct
+func (k Keeper) Hooks() Hooks {
+	return Hooks{k}
+}
+
+// Implements sdk.ValidatorHooks
+func (h Hooks) AfterValidatorBonded(ctx sdk.Context, consAddr sdk.ConsAddress, valAddr sdk.ValAddress) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/hooks.go]: Can we make the second argument `_`?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/hooks.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
+)
+
+func (k Keeper) AfterValidatorBonded(ctx sdk.Context, address sdk.ValAddress) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/hooks.go]: Do we need this hook? Can we just create `ReportInfo` on demand? That way code becomes simpler.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/app.go [hunk 5]
```diff
@@ -271,12 +265,18 @@ func NewBandApp(
 		owasm.Execute,
 		app.subspaces[oracle.ModuleName],
 		app.BankKeeper,
-		app.StakingKeeper,
+		stakingKeeper,
```
Comment (sorawit) [inline chain/app/app.go:268]: So I believe these changes are not necessary any more?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/genesis.go [hunk 6]
```diff
@@ -10,19 +10,22 @@ import (
...
 func NewGenesisState(
-	params types.Params, dataSources []types.DataSource, oracleScripts []types.OracleScript,
+	params types.Params, dataSources []types.DataSource,
+	oracleScripts []types.OracleScript, reportInfos []types.ValidatorReportInfo,
 ) GenesisState {
 	return GenesisState{
 		Params:        params,
 		DataSources:   dataSources,
 		OracleScripts: oracleScripts,
+		ReportInfos:   reportInfos,
```
Comment (sorawit) [inline chain/x/oracle/genesis.go:28]: Why is this on genesis state?

File: chain/x/oracle/keeper/infractions.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+			k.SetValidatorReportInfo(ctx, address, reportInfo)
+		}
+		logger.Info(
+			fmt.Sprintf("Validator %s missed report, but was either not found in store or already jailed", address),
+		)
+		return
+	}
+	// fetch report info
+	reportInfo, err := k.GetValidatorReportInfo(ctx, address)
+	if err != nil {
```
Comment (sorawit) [inline chain/x/oracle/keeper/infractions.go]: While it's true that here we know that the only possible error is not found, assuming `err != nil` will always be that way is antipattern. We should use `k.HasValidatorReportInfo` in my opinion. Or make `GetValidatorReportInfoWithDefault` function that returns empty report info if not exist

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/abci.go [hunk 8]
```diff
@@ -7,84 +7,84 @@ import (
...
+			if err != nil {
+				panic(err)
+
+			}
+		} else {
+			err = b.tx.Model(&Request{}).Where(Request{ID: id}).
+				Update(Request{ResolveStatus: parseResolveStatus(resolveStatus)}).Error
 			if err != nil {
 				panic(err)
+
```
Comment (Benzbeeb) [inline chain/db/abci.go]: remove newline please

File: chain/db/abci.go [hunk 9]
```diff
@@ -7,84 +7,84 @@ import (
...
+			err = b.tx.Model(&Request{}).Where(Request{ID: id}).
+				Update(
+					Request{ResolveStatus: parseResolveStatus(resolveStatus),
+						Result:      result,
+						RequestTime: requestTime,
+						ResolveTime: resolveTime,
+					}).Error
+			if err != nil {
+				panic(err)
+
```
Comment (Benzbeeb) [inline chain/db/abci.go]: ditto

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/report.go [hunk 10]
```diff
@@ -90,3 +90,21 @@ func (k Keeper) DeleteReports(ctx sdk.Context, rid types.RequestID) {
...
 	}
 }
+
+// UpdateReportInfos updates validator report info for jail validator
+// that miss report more than threshold.
+func (k Keeper) UpdateReportInfos(ctx sdk.Context, rid types.RequestID) {
+	reportedMap := make(map[string]bool)
+	iterator := k.GetReportIterator(ctx, rid)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
```
Comment (sorawit) [inline chain/x/oracle/keeper/report.go]: Should we use `k.GetReports` here?

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by Benzbeeb — APPROVED** [review]

**Pull request merged**